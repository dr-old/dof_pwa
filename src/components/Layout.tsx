import * as React from "react";
import Box from "@mui/material/Box";
import Container from "@mui/material/Container";
import { Outlet } from "react-router-dom";
import ListMenu from "./ListMenu";
import AppBar from "./AppBar";
import { styled, useMediaQuery } from "@mui/material";
import { useTheme } from "@mui/system";
import { colors } from "../utils/colors";

const BoxContent = styled(Box)(({ theme }) => ({
  bgcolor: "var(--mui-palette-background-paper)",
  display: "flex",
  position: "relative",
  minHeight: "100%",
}));

const BoxMain = styled(Box)(({ theme }) => ({
  display: "flex",
  flex: "1 1 auto",
  flexDirection: "column",
  backgroundColor:
    theme.palette.mode === "dark"
      ? colors.dark.background.darker
      : colors.light.background.darker,
}));

export default function Layout(): React.JSX.Element {
  const theme = useTheme();
  const [isMenuOpen, setIsMenuOpen] = React.useState(false); // Controls the menu slide-in/slide-out
  const isMobile = useMediaQuery(theme.breakpoints.down("sm"));

  const toggleMenu = () => {
    setIsMenuOpen((prev) => !prev);
  };

  return (
    <React.Fragment>
      <BoxContent sx={{ flexDirection: isMobile ? "column" : "row" }}>
        <BoxMain
          sx={{
            mr: { xs: 1, sm: 2, md: 0, lg: 0 },
            ml: { xs: 1, sm: 2, md: 3, lg: 4 },
            my: { xs: 1, sm: 2, md: 3, lg: 4 },
            p: { xs: 1, sm: 2, md: 3, lg: 4 },
            borderRadius: theme.shape.borderRadius * 2,
          }}>
          <AppBar
            isMenuOpen={isMenuOpen}
            toggleMenu={toggleMenu}
            isMobile={isMobile}
          />
          <main>
            <Container maxWidth="lg" sx={{ py: "64px" }}>
              <Outlet />
            </Container>
          </main>
        </BoxMain>
        <ListMenu
          isMenuOpen={isMenuOpen}
          toggleMenu={toggleMenu}
          isMobile={isMobile}
        />
      </BoxContent>
    </React.Fragment>
  );
}
