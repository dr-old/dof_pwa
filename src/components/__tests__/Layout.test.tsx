import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import Layout from "../Layout";
import { MemoryRouter } from "react-router-dom";

// Mock the AppBar and ListMenu components
jest.mock("../AppBar", () => {
  return function DummyAppBar({
    isMenuOpen,
    toggleMenu,
    isMobile,
  }: {
    isMenuOpen: boolean;
    toggleMenu: () => void;
    isMobile: boolean;
  }) {
    return (
      <div data-testid="app-bar">
        <button onClick={toggleMenu}>Toggle Menu</button>
        <span>{isMenuOpen ? "Menu Open" : "Menu Closed"}</span>
        <span>{isMobile ? "Mobile View" : "Desktop View"}</span>
      </div>
    );
  };
});

jest.mock("../ListMenu", () => {
  return function DummyListMenu({
    isMenuOpen,
    toggleMenu,
    isMobile,
  }: {
    isMenuOpen: boolean;
    toggleMenu: () => void;
    isMobile: boolean;
  }) {
    return (
      <div data-testid="list-menu">
        <span>{isMenuOpen ? "Menu Visible" : "Menu Hidden"}</span>
        <span>{isMobile ? "Mobile Menu" : "Desktop Menu"}</span>
      </div>
    );
  };
});

// Mock Outlet from react-router-dom
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  Outlet: () => <div data-testid="outlet">Outlet Content</div>,
}));

describe("Layout Component", () => {
  it("renders AppBar and ListMenu components correctly", () => {
    render(
      <MemoryRouter>
        <Layout />
      </MemoryRouter>
    );

    // Check if AppBar and ListMenu components are rendered
    expect(screen.getByTestId("app-bar")).toBeInTheDocument();
    expect(screen.getByTestId("list-menu")).toBeInTheDocument();
  });

  it("toggles the menu when the AppBar toggle button is clicked", () => {
    render(
      <MemoryRouter>
        <Layout />
      </MemoryRouter>
    );

    // Initially, the menu should be closed
    expect(screen.getByText("Menu Closed")).toBeInTheDocument();
    expect(screen.getByText("Menu Hidden")).toBeInTheDocument();

    // Click the toggle button to open the menu
    fireEvent.click(screen.getByText("Toggle Menu"));

    // The menu should now be open
    expect(screen.getByText("Menu Open")).toBeInTheDocument();
    expect(screen.getByText("Menu Visible")).toBeInTheDocument();

    // Click the toggle button to close the menu
    fireEvent.click(screen.getByText("Toggle Menu"));

    // The menu should now be closed again
    expect(screen.getByText("Menu Closed")).toBeInTheDocument();
    expect(screen.getByText("Menu Hidden")).toBeInTheDocument();
  });

  it("renders Outlet content correctly", () => {
    render(
      <MemoryRouter>
        <Layout />
      </MemoryRouter>
    );

    // Check if the Outlet content is rendered
    expect(screen.getByTestId("outlet")).toBeInTheDocument();
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <Layout />
      </MemoryRouter>
    );

    // Create a snapshot of the component
    expect(asFragment()).toMatchSnapshot();
  });
});
