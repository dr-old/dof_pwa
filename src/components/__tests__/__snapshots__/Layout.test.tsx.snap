// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-e1lug"
  >
    <div
      class="MuiBox-root css-1pm0dtq"
    >
      <div
        data-testid="app-bar"
      >
        <button>
          Toggle Menu
        </button>
        <span>
          Menu Closed
        </span>
        <span>
          Desktop View
        </span>
      </div>
      <main>
        <div
          class="MuiContainer-root MuiContainer-maxWidthLg css-qhibcd-MuiContainer-root"
        >
          <div
            data-testid="outlet"
          >
            Outlet Content
          </div>
        </div>
      </main>
    </div>
    <div
      data-testid="list-menu"
    >
      <span>
        Menu Hidden
      </span>
      <span>
        Desktop Menu
      </span>
    </div>
  </div>
</DocumentFragment>
`;
